Sources/LeafKit/LeafRenderer.swift

```diff
@@ -314,9 +314,16 @@ public final class LeafRenderer {
                     case .failure(let e): return self.eventLoop.makeFailedFuture(e)
                 }
             }
-            // return new inlined AST that may potentially still have references that can't be resolved
+            // create new AST with loaded references
             let new = LeafAST(from: ast, referencing: externals)
-            return self.cache.insert(new, on: self.eventLoop, replace: true )
+            // Check new AST's unresolved refs to see if extension introduced new refs
+            if new.unresolvedRefs.subtracting(ast.unresolvedRefs).count != 0 {
+                // AST has new references - try to resolve again recursively
+                return self.resolve(ast: new, chain: chain)
+            } else {
+                // Cache extended AST & return - AST is either flat or unresolvable
```

Comment:
Do we want to leave unresolvable ASTs here unremarked? (Assuming here that "unresolvable" means something like a circular import chain or an attempt to import a nonexistent template - in which case the whole document should be rejected as having syntax errors.) Does something notice and reject them later on, and if so, isn't caching them a waste of memory anyhow?

Reply:
What I'm trying to do is uncouple *loading and parsing* from *resolving* - so that, EG, one can arbitrarily load templates from disk at boot time and insert their *potentially* resolvable states without having to try and resolve everything. While this doesn't have a whole lot of advantages currently, where everything is assumed to be from disk, the current path for loading ASTs is SOLELY through calling render and depends on the fetch/resolve cycle to load other templates.

RE: cyclical errors, the problem is that it may be wholly dependent on another, varying syntax (eg, on an if-conditional) - any time an extension happens, you can't know if it's actually invalid in the one AST until resolution happens. In any case, before cyclical errors wouldn't error at all and just locked things up :)

RE: further on unresolvable, it may be advantageous eventually to allow other sources besides files, and again, I think it's worthwhile to get to a partially completed state regardless to avoid re-parsing the template file. If something's unresolvable, having it in the cache prevents repeated attempts to read/parse a file that otherwise would always throw unresolvable.

As it is, render, if it can't resolve, will still throw the applicable errors, but at least skips repeatedly re-reading.

This can definitely be improved, but is at worst no different than what it did before.